nomad/operator_endpoint_test.go

```diff
@@ -450,70 +472,119 @@ func (tc testcluster) leader() *Server {
 	}
 	return nil
 }
+func (tc testcluster) anyFollowerRaftServerID() raft.ServerID {
```

Comment:
This is an improvement over the old test, so no need to change it.

However I have a slight preference for just putting this code directly into the test instead of creating helper funcs. The helpers are only used once and are susceptible to logical races with "organic" leader elections caused by fun things like slow CI. This particular test would have a hard time avoiding that, but I don't want to encourage more tests assuming raft membership roles are static.

Reply:
(Adding more detail after an internal chat)

My problem with this style of helper func is that they encourage us developers to think of raft membership roles as _static_ (until we explicitly change it at least). However, leadership elections can happen at any time even during tests (slow CI builders love to trip timeouts that never happen locally).

The peerLock that is acquired prevents a _data_ race with raft membership, but not a _logical_ race. Raft can still change membership while a single peerLock is acquired. This means that the return value of these helpers might be stale or become stale by the time they're used.

For the particular test below I think that's fine. I can only think of 2 ways logical races could impact this test given a cluster of (S1: L, S2: F, S3: F):

1. anyFollower -> returns S2
2. S1 times out and S2 is elected leader "naturally"
3. TransferLeadershipTo(S2) -> Noops
4. Test passes!

So we have a false positive: the test passed but didn't really exercise the code we intended it to. Leadership was never transferred. We could detect this by looking at the Noop value of the RPC's return value and... fail? retry? Hardly seems worth it since this test is run many times and should rarely execute the incorrect code path.

The other case:

1. anyFollower -> returns S2
2. TransferLeadershipTo(S2) -> Success!
3. S2 times out and S3 is elected leader "naturally"
4. Test fails!

Oh no a false negative! Definitely bad, but the timeout would have to occur in a very small window of time. Seems incredibly unlikely, and if it's not unlikely the test will keep being flaky and we'll revisit it then. :sweat_smile: 

I can't think of a clever way to handle this false negative without a ton of effort that introduces more risks than its worth.

So the moral of the story is that these helpers cannot guarantee the freshness/accuracy of their return values. That's a pretty dangerous thing for an innocent looking helper. Few tests should care, and that's why I lean toward just hand writing the code in the 2 tests that use these ...or adding a comment or renaming the struct to newOperatorTestCluster or something to reflect its bespoke nature... just anything to avoid future us thinking these methods provide more accuracy than they actually can.